File: Python/Solution2.py
#Author: Joseph-King
#Date: Jun 21, 2022

#Returns indicies for array for values that add up to target
def twoSum(arr, target):
    #Initialize empty dictionary
    dict = {}

    #For loop to construct dictionary
    for i, num in enumerate(arr):
        #If key in dict, append index value to array
        if(num in dict):
            dict[num].append(i)
        #If key not in dict, create array with first value as index
        else:
            dict[num] = [i]

    #For loop through keys in dictionary
    for key in dict:
        #New target is target key
        new_target = target - key

        #If key is the new target, confirm there is more than 1 index in array, return indices
        if key == new_target and len(dict[key]) > 1:
            return [dict[key][0], dict[key][1]]
        #If key is not in array, and new target in dict, return indices
        elif key != new_target and new_target in dict:
            return [dict[key][0], dict[new_target][0]]


#Array test cases
arr1 = [2, 7, 11, 15]
arr2 = [3, 2, 4]
arr3 = [3, 3]

#Target test cases
target1 = 9
target2 = 6
target3 = 6

#Run Test Cases
print(twoSum(arr1, target1))
print(twoSum(arr2, target2))
print(twoSum(arr3, target3))

File: JavaScript/Solution2.js
//Author: Joseph-King
//Date: Jun 21, 2022

//Returns indicies of array for values that add up to target
function twoSum(arr, target) {

    //set map
    let map = {};

    //first for loop to convert array to hashmap
    for(let i=0; i<arr.length; i++){
        //If found push to array at key, if not create array at key
        if(map[arr[i]]){
            map[arr[i]].push(i);
        } else {
            map[arr[i]] = [i];
        }
    }
    
    //loop for each key of map
    for(let key in map){
        //determine new target
        let new_target = target - key;

        //if new target exists return array, if new target === key make sure there are two values present
        if(key === new_target && map[key].length > 1){
            return [map[key][0], map[key][1]];
        } else if (key !== new_target && map[new_target]) {
            return [map[key][0], map[new_target][0]];
        }
    }
}

//Array test cases
var arr1 = [2, 7, 11, 15];
var arr2 = [3, 2, 4];
var arr3 = [3, 3];

//Target test cases
var target1 = 9;
var target2 = 6;
var target3 = 6;

//Run Test Cases
console.log(twoSum(arr1, target1));
console.log(twoSum(arr2, target2));
console.log(twoSum(arr3, target3));